### PR TITLE
Add explicit implicits

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by rorygraves** [comment]
Cool stuff, but drone build is going to fail:
@dwijnand There are failures - http://fommil.com/ensime/ensime-server/381

```
[error] /drone/src/github.com/ensime/ensime-server/protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala:46: ambiguous implicit values:
[error]  both object SymbolJsonFormat in trait BasicFormats of type org.ensime.jerk.JerkConversions.SymbolJsonFormat.type
[error]  and value symbolFormat in object JerkConversions of type => spray.json.JsonFormat[Symbol]
[error]  match expected type spray.json.JsonFormat[Symbol]
[error]   val EnsimeServerMessageFormat: RootJsonFormat[EnsimeServerMessage] = cachedImplicit
[error]       ^
```

**Comment by dwijnand** [comment]
Given the types of errors (NPEs -.-) and the number of changes, I think the way forward here is keep the current diff as the target, and rebuild the changes incrementally, testing along the way (unless someone can figure out what broke the build). It'll take me some time, so feel free to close this if it's in the way.

**Comment by fommil** [comment]
I'll have a look, just need to find the time. My super high top priority is https://github.com/fommil/class-monkey so expect me to go to ground until it's finished. Maybe @rorygraves can find some time to have a quick look?

**Comment by fommil** [comment]
but yes, I agree that if you can get as much stuff into an initial PR as possible that would be good, then we can investigate what is causing the problem and beg Miles to help.

**Comment by fommil** [comment]
how very bizarre. I'm seeing a reasonably minimal failure with this

```
s-express/testOnly *BasicFormatsSpec -- -z Long
```

which should help localise the issue. I don't think shapeless can even be blamed for this one.

**Comment by fommil** [comment]
@dwijnand woohoo! Fixed it! I'll send you a PR that you can just merge into here or cherry pick. I'm happy to squash everything into one big commit with your name on it.

**Comment by dwijnand** [comment]
Cool. Thanks for last mileing it so it can go in.

**Comment by fommil** [comment]
@dwijnand well, **almost** fixed it. Still getting problems in the SwankFormatsSpec but I think I know what the problem is now.

**Comment by fommil** [comment]
@dwijnand can you please cherry pick the commit from https://github.com/fommil/ensime-server/commits/explicit-implicits ?

I just ran the unit tests, still getting a failure, and it's probably to do with my quick and dirty `lazyImplicit` method entering into infinite recursion. It really needs the equivalent of the `cachedImplicit` macro but for `Lazy`.

**Comment by fommil** [comment]
asked here about the `lazyImplicit` https://gitter.im/milessabin/shapeless?at=57100bce5cd40114649b654c I think it's bedtime for me.

File: protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala [hunk 1]
```diff
@@ -19,7 +19,7 @@ private object JerkConversions extends DefaultJsonProtocol with FamilyFormats {
   implicit override def eitherFormat[A: JsonFormat, B: JsonFormat]: JsonFormat[Either[A, B]] = super.eitherFormat[A, B]
   // Note that its not possible to override an object in scala, so we
   // just define a new one that wins the race.
-  implicit val symbolFormat = SymbolJsonFormat
+  implicit val symbolFormat: SymbolJsonFormat.type = SymbolJsonFormat
```
Comment (fommil) [inline protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala]: this is so weird

Comment (dwijnand) [inline protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala]: Weird-wrong-looking or weird-bizarre?

Comment (fommil) [inline protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala]: Weird bizarre. I understand why it's needed though but I wouldn't have realised it before you added it.

Comment (dwijnand) [inline protocol-jerky/src/main/scala/org/ensime/jerk/JerkFormats.scala]: It's the most specific type possible. I'm not sure it's what is normally inferred, but this seemed to do the trick.

**Comment by fommil** [comment]
I'm surprised the CI is going green, I got red locally.

**Comment by fommil** [comment]
I think my laptop is just not fast enough to run the 8 tests in parallel :smile: this is great, thanks @dwijnand !

**Comment by dwijnand** [comment]
Oh sweet. Thanks again for helping me get this done.

**Comment by fommil** [comment]
can we expect more rules enforced with scapegoat or whatever it is you used? You might also like #1148

**Comment by dwijnand** [comment]
Yeah I'll probably cycle (no pun intended) back to these kind of improvements over time :)

**Pull request merged**